### PR TITLE
Docs: Add docs to `RepartitionExec` and architecture guide

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

N/A
# Rationale for this change

While writing up https://github.com/apache/arrow-datafusion/issues/7001 I made some diagrams and descriptions about `Repartition` that I felt would add value as doc comments

# What changes are included in this PR?
Add docs to `RepartitionExec` and architecture guide

# Are these changes tested?
CI tests the doc comments

# Are there any user-facing changes?
Docs

### CODE REVIEW

File: datafusion/core/src/physical_plan/repartition/mod.rs [hunk 1]
```diff
@@ -239,8 +239,67 @@ impl BatchPartitioner {
     }
 }
 
-/// The repartition operator maps N input partitions to M output partitions based on a
-/// partitioning scheme. No guarantees are made about the order of the resulting partitions.
+/// Maps `N` input partitions to `M output partitions based on a
```
Comment (alamb) [inline datafusion/core/src/physical_plan/repartition/mod.rs:242]: Screenshot 2023-07-17 at 12 21 43 PM

**Review by r4ntix — APPROVED** [review]
❤️

File: datafusion/core/src/physical_plan/repartition/mod.rs (hunk 1 above)
Comment (2010YOUY01) [inline datafusion/core/src/physical_plan/repartition/mod.rs:242]:
```suggestion
/// Maps `N` input partitions to `M` output partitions based on a
```

**Review by 2010YOUY01 — COMMENTED** [review]
Really nice doc 👍🏼

**Pull request merged**